---
  title: "Coder Upgrade 2023: Multiomics analysis"
author: "Matt Cannon"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output:
  html_document:
  toc: true
toc_float: true
toc_depth: 5
number_sections: false
code_folding: show
---
  
  ```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      cache = TRUE,
                      cache.lazy = FALSE)
```

Load libraries
```{r libraries, cache=FALSE, warning=FALSE, error=FALSE, message=FALSE}
library(rrrSingleCellUtils)
library(tidyverse)
library(Seurat)
library(Signac)
library(patchwork)
system("ml HDF5")
```

--------------------------------------------------------------------------------
  
  # Section 1: Overview
  
  ## What is multiomics single-cell data?
  
  Any technology that produces multiple type of data from single cells

Each cell is profiled for multiple types of data

- RNA / ATAC
- RNA / ChIP
- RNA / Protein
- RNA / DNA
- RNA / DNA methylation / Chrom. accessibility

## Benefits of multiomics

- Direct association within individual cells between different types of data
- Avoids shakey correlations made using bulk methods on heterogeneous populations
- Can generate insights not possible with other methods

## Challenges of multiomics

- Generally, the data quality will be much worse for each individual cell compared to bulk methods
- Sample/library preparation is more complex
- Especially for homebrew methods
- $$$
  - Analysis is going to be much more complex
- Need to analyze multiple single-cell data types
- Need proper statistical methods to integrate the data
- Methods are generally less mature than more commonly used methods

## 10x multiomics kit

- Generates both RNA and ATAC data from single nuclei
- Relies on cleanly isolated and fixed nuclei
- Get only RNAs from the nucleus
- Nuclei prep can be quite challenging
- Likely more prone to cell-cell cross contamination

## 10x multiomics data requirements

RNA and ATAC have different sequencing requirements

- RNA
- R1: 28bp <- has UMI and nucleus barcode
- R2: 90bp <- has RNA sequence
- I1/I2: 10bp <- have sample barcodes
- ATAC
- R1: 50bp <- has genomic sequence
- R2: 24bp <- has nucleus barcode
- R3: 49bp <- has genomic sequence
- I1: 8bp <- has sample barcode
- Recommended 20,000 read pairs / nucleus for GEX
- In 10x demo dataset, 122,335 reads per nucleus
- Recommended 25,000 read pairs / nucleus for ATAC
- In 10x demo datasets, 95,897 reads per nucleus
- Aim higher if possible
- For 2000 cells, 100k + 100k reads per sample
- 400M reads per sample
- ~4 samples on a NovaSeq S1

## Data

Depending on what data type you recieve you may need to do different things

- BCLs
- Run cellranger-arc mkfastq
- If both GEX and ATAC run on single lane, need to specify read lengths for each
- If run seperately on different lanes, can just run mkfastq on each lane
- Fastqs
- Run cellranger-arc count
- Simple way is to do one run per sample
- Input is csv with sample info
- Output is a folder with a bunch of files/folders
- https://support.10xgenomics.com/single-cell-multiome-atac-gex/software/pipelines/latest/using/count
- web_summary.html              - html report
- filtered_feature_bc_matrix/   - contains the GEX data
- filtered_feature_bc_matrix.h5 - h5 file with both GEX and ATAC data
- atac_peaks.bed                - ATAC peaks
- atac_fragments.tsv.gz         - ATAC fragments
- cloupe.cloupe                 - Loupe browser file
- *.bam                         - BAM files

--------------------------------------------------------------------------------
  
  ## Activity 1 (1:30 - 2ish)
  
  Talk to Amy Gross, Meren Cam and Jason Navarro

--------------------------------------------------------------------------------
  
  ## Using Seurat and Signac for 10x multiomics data individually
  
  You essentially get two independant datasets back (GEX and ATAC) that can be analyzed independently.

The GEX data can be analyzed like normal

- One exception - if you read in your data using `Read10X_h5()`, it returns a list of two objects:
  - Gene Expression
- Peaks
- Need to pull the GEX data out of the list and use that for analysis

I'm not going to walk through the GEX analysis as that's been covered in previous sessions
```{r}
data_path <- "/home/gdworkshop/lab/session_data/08/"
# If you're running the code on the IGM AWS instance, you can use this path:
#data_path = "/igm/projects/scrgot_coder_upgrade/2023/"

h5_data <-
  Read10X_h5(paste0(data_path,
                    "e18_mouse_brain_fresh_5k_filtered_feature_bc_matrix.h5"))

# mt_pattern <- "^mt-" because this is mouse data
seurat_obj <-
  CreateSeuratObject(counts = h5_data$`Gene Expression`,
                     assay = "RNA",
                     project = "10x_multiomics",
                     min_cells = 5,
                     min_features = 800) %>%
  PercentageFeatureSet(pattern = "^mt-",
                       col.name = "percent_mt_rna",
                       assay = "RNA")
```

## Add the ATAC data to the Seurat object
Need to combine the datasets prior to filtering
```{r}
frag_file <-
  paste0(data_path,
         "e18_mouse_brain_fresh_5k_atac_fragments.tsv.gz")

seurat_obj[["ATAC"]] <-
  CreateChromatinAssay(counts = h5_data$Peaks,
                       sep = c(":", "-"),
                       fragments = frag_file,
                       min.cells = 5)
```

# Filter out likely junk cells
Not going to walk through this as it should all be review
```{r}
cutoffs <-
  tribble(~feature, ~min_val, ~max_val,
          "nCount_RNA", 2000, 25000,
          "nFeature_RNA", 1000, 7000,
          "percent_mt_rna", 0, 20,
          "nCount_ATAC", 2000, 60000,
          "nFeature_ATAC", 1000, 25000)

feature_hist(seurat_obj,
             features = c("nCount_RNA",
                          "nFeature_RNA",
                          "percent_mt_rna",
                          "nCount_ATAC",
                          "nFeature_ATAC"),
             cutoff_table = cutoffs)

length(Cells(seurat_obj))
seurat_obj <-
  subset(seurat_obj,
         subset = nCount_RNA >= cutoffs$min_val[1] &
           nCount_RNA <= cutoffs$max_val[1] &
           nFeature_RNA >= cutoffs$min_val[2] &
           nFeature_RNA <= cutoffs$max_val[2] &
           percent_mt_rna <= cutoffs$max_val[3]) %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA() %>%
  RunUMAP(reduction.name = "umap_rna",
          dims = 1:10) %>%
  FindNeighbors(dims = 1:10) %>%
  FindClusters(resolution = 0.4)

seurat_obj$RNA_cluster <-
  Idents(seurat_obj)

length(Cells(seurat_obj))

# Not the best way to characterize brain cell types, but it's good enough for now
imm_cells <- celldex::ImmGenData()
mouse_rna <- celldex::MouseRNAseqData()

cell_assign <-
  SingleR::SingleR(as.SingleCellExperiment(seurat_obj),
                   ref = list(imm_cells,
                              mouse_rna),
                   labels = list(imm_cells$label.fine,
                                 mouse_rna$label.fine))

seurat_obj$cell_type <-
  cell_assign$labels %>%
  str_replace("DC", "Dendritic cells")
seurat_obj$cell_score <-
  cell_assign$scores %>%
  apply(MARGIN = 1, function(x) max(x, na.rm = TRUE))

DimPlot(seurat_obj,
        group.by = "cell_type",
        reduction = "umap_rna",
        label = TRUE)

# Save this so I don't have to re-run everything
qs::qsave(seurat_obj, paste0(data_path,
                             "mouse_brain_5k_sobj_1.qs"))
```

# Let's focus on the ATAC data

## Add in gene annotations
```{r}
DefaultAssay(seurat_obj) <- "ATAC"

annotations <-
  GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79::EnsDb.Mmusculus.v79)
seqlevelsStyle(annotations) <- 'UCSC'
genome(annotations) <- "mm10"

# add the gene information to the object
Annotation(seurat_obj[["ATAC"]]) <- annotations
```

## Add in nucleosome signal.
This is a the ratio of mononucleosome fragments to short fragments.
```{r}
seurat_obj <- NucleosomeSignal(seurat_obj)
seurat_obj$nucleosome_group <-
  ifelse(seurat_obj$nucleosome_signal > 4,
         paste0("NS > ", 4),
         paste0("NS < ", 4))
```

## Calculate the signal enrichment at transcription start sites
```{r}
seurat_obj <- TSSEnrichment(seurat_obj, fast = FALSE)
seurat_obj$high_tss <-
  ifelse(seurat_obj$TSS.enrichment > 2,
         "High",
         "Low")
```

## Calculate how many of the reads are in different types of regions
```{r}
total_frag_df <-
  CountFragments(paste0(data_path,
                        "e18_mouse_brain_fresh_5k_atac_fragments.tsv.gz"))

total_frag_df <-
  total_frag_df %>%
  dplyr::filter(CB %in% colnames(seurat_obj)) %>%
  dplyr::arrange(match(CB, colnames(seurat_obj)))

# Populate seurat_obj with metadata

seurat_obj$total_frag <- total_frag_df$reads_count
seurat_obj$mononucleosomal <- total_frag_df$mononucleosomal
seurat_obj$nucleosome_free <- total_frag_df$nucleosome_free
```

## Calculate the fraction of reads in peaks
```{r}
seurat_obj <-
  FRiP(seurat_obj,
       assay = "ATAC",
       total.fragments = "total_frag",
       col.name = "FRiP",
       verbose = FALSE)
```

## Add in metadata output by cellranger-arc count
```{r}
metadata <-
  read_csv(paste0(data_path,
                  "/e18_mouse_brain_fresh_5k_per_barcode_metrics.csv"),
           show_col_types = FALSE) %>%
  mutate(atac_pct_mito = atac_mitochondrial_reads / atac_raw_reads * 100,
         atac_pct_dups = atac_dup_reads / atac_raw_reads * 100) %>%
  filter(barcode %in% colnames(seurat_obj)) %>%
  select(-atac_barcode,
         -is_cell,
         -excluded_reason,
         -gex_barcode) %>%
  arrange(match(barcode, colnames(seurat_obj))) %>%
  column_to_rownames("barcode")

seurat_obj <- AddMetaData(seurat_obj, metadata)

# Save it so I don't have to re-run everything
qs::qsave(seurat_obj, paste0(data_path,
                             "mouse_brain_5k_sobj_2.qs"))
```

## Lets plot our new metadata
```{r}
plot_factors <-
  c("TSS.enrichment",
    "FRiP",
    "nucleosome_signal",
    "atac_pct_mito",
    "nCount_ATAC",
    "atac_pct_dups")
plot_name <- feature_hist(seurat_obj, plot_factors)
ggsave("Sessions/08/atac_feat_hist.png",
       plot_name,
       width = 6,
       height = 10)
```

Generally want:
  
  -   TSS enrichment > 1
-   FRiP > 0.2
-   Nucleosome signal < 4

Also:
  You're going to have high PCR duplicate rates unless your sequencing depth is really shallow. Not a big deal. The number will give you an idea of how "rich" your library was.

High ATAC percent mito should be low, higher percents may indicate "dirty" nuclei

## Process the seurat data
For ATAC, we do a couple of things differentially

- We don't use the standard normalization, we use term frequency-inverse document frequency (TF-IDF) instead
- "This is a two-step normalization procedure, that both normalizes across cells to correct for differences in cellular sequencing depth, and across peaks to give higher values to more rare peaks."
- We don't run PCA
    - We run singular value decomposition (SVD) on the TD-IDF matrix generated by RunTDIDF()
    - "In linear algebra, the singular value decomposition (SVD) is a factorization of a real or complex matrix. It generalizes the eigendecomposition of a square normal matrix with an orthonormal eigenbasis to any m x n matrix." - Wikipedia
    - "The combined steps of TF-IDF followed by SVD are known as latent semantic indexing (LSI), and were first introduced for the analysis of scATAC-seq data by Cusanovich et al. 2015."
    - The reduction is named "lsi"

```{r}
seurat_obj <-
    seurat_obj %>%
    RunTFIDF() %>%
    FindTopFeatures(min.cutoff = "q0") %>%
    RunSVD() %>%
    FindNeighbors(reduction = "lsi") %>%
    FindClusters(algorithm = 3) %>%
    RunUMAP(reduction = "lsi",
            dims = 2:30,
            reduction.name = "umap_atac")

seurat_obj$ATAC_cluster <- Idents(seurat_obj)

DimPlot(seurat_obj,
        group.by = "cell_type",
        reduction = "umap_atac",
        label = TRUE)

qs::qsave(seurat_obj, paste0(data_path,
                             "mouse_brain_5k_sobj_3.qs"))
```

### Find differential accessibility between cell types
Also add in annotation for nearest feature
```{r}
Idents(seurat_obj) <- "cell_type"

da_peaks <-
    FindMarkers(seurat_obj,
                ident.1 = "NPCs",
                ident.2 = "Neurons",
                test.use = "LR",
                latent.vars = "atac_peak_region_fragments") %>%
    rownames_to_column("region") %>%
    left_join(ClosestFeature(seurat_obj,
                             regions = .$region) %>%
        dplyr::rename(region = query_region))
```

# Plot ATAC peaks
Let's pick out three up and three down regulated regions
```{r}
fav_peaks <-
  da_peaks %>%
  filter(p_val_adj < 1e-20 &
           abs(avg_log2FC) > 0.5 &
           gene_biotype == "protein_coding" &
           distance == 0) %>%
  group_by(avg_log2FC > 0) %>%
  arrange(avg_log2FC) %>%
  slice_head(n = 3) %>%
  pull(region)

CoveragePlot(seurat_obj,
             region = fav_peaks,
             idents = c("NPCs", "Neurons"),
             extend.upstream = 2000,
             extend.downstream = 2000)
```

One issue:
  
  If you have multiple samples called with cellranger-arc, their peaks are different, so you can't really compare them directly.
I'm not going to walk through this, due to time constraints, but just wanted to show a couple of options to deal with this.

You can either re-call the peaks: https://stuartlab.org/signac/articles/peak_calling.html

Or redefine their boundaries: https://stuartlab.org/signac/articles/merging.html

https://github.com/kidcancerlab/rrrSingleCellUtils/blob/master/R/multiomics.R#L18

This code wont work since the files dont exist, its just to show you how to use the function
```{r eval=FALSE}
# Documentation is still a work in progress

library(rrrSingleCellUtils)
peak_beds <-
    paste0(data_path,
           c("sample_1/atac_peaks.bed",
             "sample_2/atac_peaks.bed",
             "sample_3/atac_peaks.bed"))

frag_paths <-
    paste0(data_path,
           c("sample_1/atac_fragments.tsv.gz",
             "sample_2/atac_fragments.tsv.gz",
             "sample_3/atac_fragments.tsv.gz"))

cell_ids <-
    c("sample_1",
      "sample_2",
      "sample_3")

# Uses {parallel} to run faster
merged_atac_data <-
    merge_atac(peak_beds = peak_beds,
               frag_paths = frag_paths,
               cell_ids = cell_ids,
               n_regions_simul = 100000,
               threads = 3) %>%
    subset(nCount_ATAC > 100)
```

--------------------------------------------------------------------------------

## Activity 2 (2:30 - 3ish)

Quiz:

1. What is my name? Matthew Vesuvius Cannon
2. What is my quest? To vanquish lines greater than 80 characters
3. What is the airspeed velocity of an unladen swallow? 42
4. T/F: for 10X multiomics data, both the RNA and ATAC data have identical read lengths FALSE
5. How many reads per cell should you target for 10X multiome ATAC data? 100k
6. T/F: for 10X multiome ATAC data, we count number of peaks per gene to calculate gene activity TRUE
7. After reading in the GEX data, how many cells are in my Seurat object? 
8. I filter bad cells after I merge the GEX and ATAC datasets. What problems might I run into if I filter bad cells before merging?
9. How many cells do I lose when I filter out bad cells?
10. T/F: Single-cell and single-nuclei GEX data would be quite different if you could magically run them both on the same cells
11. What kind of dimensional reduction technique do we run on the ATAC data?
12. Why can't you easily merge ATAC data from multiple samples?
  13. How many peaks are in the ATAC dataset?
  --------------------------------------------------------------------------------
  ## Joint analysis of multimodal data
  
  > Show ppt of plots

### Matching cells between datasets
Compare heterogeneity of gene expression vs. chromatin accessibility

```{r}
seurat_obj <-
  qs::qread(paste0(data_path,
                   "mouse_brain_5k_sobj_3.qs"))

rna_umap <-
  Embeddings(seurat_obj, reduction = "umap_rna") %>%
  as.data.frame() %>%
  rownames_to_column("cell") %>%
  as_tibble() %>%
  mutate(rna_cluster = as.vector(seurat_obj$RNA_cluster))

atac_umap <-
  Embeddings(seurat_obj, reduction = "umap_atac") %>%
  as.data.frame() %>%
  rownames_to_column("cell") %>%
  as_tibble() %>%
  dplyr::rename(UMAP_1 = umap_atac_1,
                UMAP_2 = umap_atac_2) %>%
  full_join(rna_umap %>%
              select(cell, rna_cluster))

# Merging the UMAP data
# Shifted UMAP_1 for both by offset_num to separate the datasets

offset_num <- 13

combined_umap <-
  rna_umap %>%
  mutate(data_type = "RNAseq",
         UMAP_1 = UMAP_1 - offset_num) %>%
  bind_rows(atac_umap %>%
              mutate(data_type = "ATACseq",
                     UMAP_1 = UMAP_1 + offset_num))

draw_cluster <- 7

# Make plot of the combined UMAP projections
ggplot(combined_umap,
       aes(x = UMAP_1,
           y = UMAP_2,
           shape = data_type)) +
  geom_point(aes(color = rna_cluster == draw_cluster),
             alpha = 0.5) +
  scale_color_brewer(palette = "Paired",
                     name = "Data type") +
  geom_line(data = combined_umap %>%
              filter(rna_cluster == draw_cluster),
            aes(x = UMAP_1,
                y = UMAP_2,
                group = cell),
            color = "#808080",
            alpha = 0.05) +
  scale_color_brewer(palette = "Set2",
                     name = "Target cluster")

ggsave("Sessions/08/combined_umap.png",
       width = 10,
       height = 6)
```

### Multimodal UMAP
Better separation of cell types due to increased dimensionality of data
```{r}
seurat_obj <-
  seurat_obj %>%
  FindMultiModalNeighbors(reduction.list = list("pca", "lsi"),
                          dims.list = list(1:20, 2:20))

# build a joint UMAP visualization
seurat_obj <-
  seurat_obj %>%
  RunUMAP(nn.name = "weighted.nn",
          assay = "RNA",
          reduction.name = "joint_umap") %>%
  FindClusters(algorithm = 3,
               graph.name = "wsnn")

seurat_obj$joint_cluster <-
  Idents(seurat_obj)

joint_umap <-
  DimPlot(seurat_obj,
          reduction = "joint_umap",
          label = TRUE,
          repel = TRUE,
          pt.size = 0.5,
          group.by = "cell_type") +
  NoLegend()

rna_umap <-
  DimPlot(seurat_obj,
          reduction = "umap_rna",
          label = TRUE,
          repel = TRUE,
          pt.size = 0.5,
          group.by = "cell_type") +
  NoLegend()

atac_umap <-
  DimPlot(seurat_obj,
          reduction = "umap_atac",
          label = TRUE,
          repel = TRUE,
          pt.size = 0.5,
          group.by = "cell_type") +
  NoLegend()

joint_umap + (rna_umap / atac_umap)
```

### Correlation of ATAC peaks with gene expression
```{r}
DefaultAssay(seurat_obj) <- "ATAC"
Idents(seurat_obj) <- "cell_type"

seurat_obj <-
  RegionStats(seurat_obj,
              genome = BSgenome.Mmusculus.UCSC.mm10::BSgenome.Mmusculus.UCSC.mm10)
# ignore the warning - our object just doesn't have some of the junk contigs

diff_peaks <-
  FindMarkers(seurat_obj,
              assay = "ATAC",
              ident.1 = "Neurons",
              ident.2 = "NPCs") %>%
  rownames_to_column("region") %>%
  as_tibble() %>%
  full_join(tibble(region = rownames(seurat_obj))) %>%
  arrange(match(region, rownames(seurat_obj))) %>%
  mutate(p_val_adj = ifelse(is.na(p_val_adj), 1, p_val_adj))

qs::qsave(diff_peaks,
          "Sessions/08/diff_peaks.qs")

seurat_obj[["ATAC"]][["neuron_signif"]] <- diff_peaks$p_val_adj < 0.05

de_genes <-
  FindMarkers(seurat_obj,
              assay = "RNA",
              ident.1 = "Neurons",
              ident.2 = "NPCs")

# link peaks to genes
seurat_obj <-
  LinkPeaks(object = seurat_obj,
            peak.assay = "ATAC",
            expression.assay = "RNA",
            min.distance = 5000,
            genes.use = rownames(de_genes)[1:20])

links <-
  Links(seurat_obj) %>%
  as.data.frame() %>%
  as_tibble() %>%
  arrange(pvalue)

links

CoveragePlot(object = seurat_obj,
             region = "Pax6",
             features = "Pax6",
             idents = c("Neurons", "NPCs"),
             peaks.group.by = "neuron_signif",
             expression.assay = "RNA",
             extend.upstream = 150000,
             extend.downstream = 50000)
```

## Motif analysis of peak sequence can potentially identify binding factors
```{r eval=FALSE}
freq_matrix <-
  TFBSTools::getMatrixSet(x = JASPAR2020::JASPAR2020,
                          opts = list(collection = "CORE",
                                      tax_group = "vertebrates",
                                      all_versions = FALSE))

# The chromosomes for this dataset are a bit goofy so we need to kick out a couple junk chromosomes
seurat_obj <-
  seurat_obj[grep("^GL|^JH", rownames(seurat_obj), invert = TRUE), ]

# add motif information
# Ignore the warning - our object just doesn't have some of the junk contigs
seurat_obj <-
  AddMotifs(object = seurat_obj,
            genome = BSgenome.Mmusculus.UCSC.mm10::BSgenome.Mmusculus.UCSC.mm10,
            pfm = freq_matrix)

Idents(seurat_obj) <- "cell_type"

diff_peaks <-
  FindMarkers(object = seurat_obj,
              ident.1 = "Neurons",
              ident.2 = "NPCs",
              only.pos = TRUE,
              test.use = "LR",
              min.pct = 0.1,
              latent.vars = "nCount_ATAC")

top_diff_peaks <-
  rownames(diff_peaks[diff_peaks$p_val < 0.005, ])

open_peaks <-
  AccessiblePeaks(seurat_obj,
                  idents = c("Neurons", "NPCs"))

# match the overall GC content in the peak set
meta_feature <-
  GetAssayData(seurat_obj,
               assay = "ATAC",
               slot = "meta.features")

peaks_matched <-
  MatchRegionStats(meta.feature = meta_feature[open_peaks, ],
                   query.feature = meta_feature[top_diff_peaks, ],
                   n = 50000)

## Matching GC.percent distribution

# test enrichment
enriched_motifs <-
  FindMotifs(object = seurat_obj,
             features = top_diff_peaks)

head(enriched_motifs)

motif_locs <-
  Motifs(seurat_obj)@positions %>%
  as.data.frame() %>%
  full_join(tibble(motif_name = Motifs(seurat_obj)@motif.names %>%
                     as.character(),
                   group_name = Motifs(seurat_obj)@motif.names %>%
                     names())) %>%
  as_tibble()

motif_locs
```

--------------------------------------------------------------------------------
  
  ## Activity 3 (3:30 - 4ish)
  
  Quiz:
  
  1. Why will a UMAP performed using both RNA and ATAC data be more informative than a UMAP performed using only one dataset? Increased dimensionality
2. Why might the cells from a cluster in one of the datasets be randomly scattered in the other dataset? Microenvironment/convergence
3. Why might we be interested in an ATAC peak that is correlated with the expression of a gene but is not located in the promoter region of (or even close to) that gene? enhancer/superenhancer
4. What function creates a weighted nearest neighbor graph? 
  5. What function finds correlations between gene expression and ATAC peaks?
  6. T/F: Correlation is equivalent to causation.
7. T/F: Within a single nuclei you cannot find a correlation between human gene expression patterns and mouse ATAC peaks.
8. Why might single-nuclei single-cell sequencing be more prone to cross contamination than single-cell sequencing of dissociated cells?
  
  --------------------------------------------------------------------------------
  
  ## Session challenge
  
  10X has a multiomics dataset that is from a lymph node with B cell lymphoma. The dataset can be found here:
  https://www.10xgenomics.com/resources/datasets/fresh-frozen-lymph-node-with-b-cell-lymphoma-14-k-sorted-nuclei-1-standard-2-0-0

A summary of the data and the analysis can be found here:
  https://pages.10xgenomics.com/rs/446-PBO-704/images/10x_LIT000110_Data_Spotlight_Multiome_digital.pdf

Download this dataset from 10X and find differential expression between the tumor cells and the normal B cells. Also find differential chromatin accessibility for the same comparison. Then pick a few top differentially expressed genes that look interesting biologically and find nearby ATAC peaks that correlate with the expression of those genes.

Bonus points if you can also find motifs that are located in the ATAC peaks that are both differentially accessible and significantly correlated with gene expression.

--------------------------------------------------------------------------------
  
  # Resources
  Methods and applications for single-cell and spatial multi-omics: https://www.nature.com/articles/s41576-023-00580-2

Into the multiverse: advances in single-cell multiomic profiling: https://www.cell.com/trends/genetics/fulltext/S0168-9525(22)00077-4

https://www.10xgenomics.com/products/single-cell-multiome-atac-plus-gene-expression

https://stuartlab.org/signac/
  
  Merging ATAC datasets: https://stuartlab.org/signac/articles/merging.html